src/core/dyad_core.c

```diff
@@ -0,0 +1,622 @@
+            free((*ctx)->kvs_namespace);
+            free((*ctx)->prod_managed_path);
+            free(*ctx);
+            *ctx = NULL;
+            return DYAD_NOCTX;
+        }
+        strncpy(
+                (*ctx)->cons_managed_path,
+                cons_managed_path,
+                strlen(cons_managed_path)+1
+            );
+    }
+    if ((*ctx)->prod_managed_path == NULL && (*ctx)->cons_managed_path == NULL)
+    {
+        fprintf (stderr, "Warning: DYAD instance is neither producer nor consumer!\n");
+    }
+    printf("Opening Flux\n");
+    (*ctx)->reenter = true;
+    (*ctx)->h = flux_open(NULL, 0);
+    if ((*ctx)->h == NULL)
+    {
+        fprintf (stderr, "Could not open Flux handle!\n");
+        return DYAD_FLUXFAIL;
+    }
+    printf("Getting Flux Rank\n");
+    if (flux_get_rank((*ctx)->h, &((*ctx)->rank)) < 0)
+    {
+        fprintf(stderr, "Could not get Flux rank\n");
+        FLUX_LOG_ERR ((*ctx)->h, "Could not get Flux rank!\n");
+        return DYAD_FLUXFAIL;
+    }
+    printf("Initialization complete\n");
+    (*ctx)->initialized = true;
+    // TODO Print logging info
+    return DYAD_OK;
+}
+
+int dyad_produce(dyad_ctx_t *ctx, const char *fname)
+{
+    int rc = validate_context(ctx, false, true);
+    if (DYAD_IS_ERROR(rc))
+    {
+        return rc;
+    }
+    return dyad_commit(ctx, fname);
+}
+
+int dyad_consume(dyad_ctx_t *ctx, const char *fname)
+{
+    int rc = validate_context(ctx, true, false);
+    if (DYAD_IS_ERROR(rc))
+    {
+        return rc;
+    }
+    dyad_kvs_response_t *resp = NULL;
+    rc = dyad_fetch(ctx, fname, &resp);
+    if (rc != 0)
+    {
+        DYAD_LOG_ERR (ctx, "dyad_fetch failed!\n");
+        if (resp != NULL)
+        {
+            dyad_free_kvs_response(resp);
+        }
+        return rc;
+    }
+    rc = dyad_pull(ctx, fname, resp);
+    dyad_free_kvs_response(resp);
+    if (rc != 0)
+    {
+        DYAD_LOG_ERR (ctx, "dyad_free_kvs_response failed!\n");
+        return rc;
+    };
+    return DYAD_OK;
+}
+
+#if DYAD_PERFFLOW
+__attribute__((annotate("@critical_path()")))
+#endif
+int dyad_kvs_commit(dyad_ctx_t *ctx, flux_kvs_txn_t *txn,
+        const char *upath, flux_future_t **f)
+{
+    printf("Committing data to KVS\n");
+    *f = flux_kvs_commit(ctx->h, ctx->kvs_namespace,
+            0, txn);
+    if (*f == NULL)
+    {
+        DYAD_LOG_ERR(ctx, "Could not commit transaction to Flux KVS\n");
+        return DYAD_BADCOMMIT;
+    }
+    printf("Wait for commit to complete\n");
+    flux_future_wait_for(*f, -1.0);
+    printf("Destroy future for KVS commit\n");
+    flux_future_destroy(*f);
+    *f = NULL;
+    printf("Destroy transaction on KVS\n");
+    flux_kvs_txn_destroy(txn);
+    return DYAD_OK;
+}
+
+#if DYAD_PERFFLOW
+__attribute__((annotate("@critical_path()")))
+#endif
+int publish_via_flux(dyad_ctx_t* ctx, const char *upath)
+{
+    printf("Initial publication setup\n");
+    const char *prod_managed_path = ctx->prod_managed_path;
```

Comment:
I suspect that this line would cause a compiler warning with certain C compilers. There are some that expects the variable initialization with a value that is known at compile time. So, to avoid such a warning, the pointer can be set to NULL at variable definition, and then later set to the value after all the local variables are initialized.